README.md
# NAME

tap - add a tap method everywhere

# VERSION

version 0.01

# SYNOPSIS

    use tap;

    return Some::Class->new
      ->tap::color('red')
      ->tap::smell('roses');

    # rather than:

    my $o = Some::Class->new;
    $o->color('red')
    $o->smell('roses');
    return $o;

# DESCRIPTION

`tap` is a method that wraps any other method, ignoring its result and
returning the invocant instead.  This allows you to make chaining method out of
anything, such a setters that return what you just set, or to peek into
a method chain.

This module gets you access to tap anywhere, without requiring it to exist in
a parent class or an imported lexical.

# METHODS

## tap::\*

    $o->tap::color('red')
      ->tap::smell('roses')
      ->...;

For use when you know what existing method you want to call.  This is
implemented using AUTOLOAD, which follows the common behavior of installing
a method so you only fall to AUTOLOAD once.

It's not as fast as writing things out the long way, but it's the fastest of
all tap-like features I've seen.  And I think it's convenient while looking
nice.  I was using it in [Reply](https://metacpan.org/pod/Reply) for quite a while before making it a module.

## tap::tap

    $o->tap::tap($method => 42)->...;

    # slip debugging into a chain
    use DDP;
    $o->tap::tap(\&p)->run;

For when the method name is dynamic, or you want to pass a coderef to be called
as a method.  For convenience the invocant is topicalized (`$_`) as well:

    $o->tap::tap(sub { $_->foo->bar(42) })->run;

# SEE ALSO

[Object::Tap](https://metacpan.org/pod/Object::Tap), [Object::Util](https://metacpan.org/pod/Object::Util), [Mojo::Base](https://metacpan.org/pod/Mojo::Base) - other taps

[curry](https://metacpan.org/pod/curry) - interface inspiration

# PERFORMANCE

The following example benchmark (`ex/bench.pl`) was run with perl 5.26.0 on
a 2015 MacBook Pro.

                      Rate tap-tap Object::Tap tap-reftap own-tap tap-autoload no-tap
    tap-tap       650759/s      --         -1%        -5%    -14%         -30%   -49%
    Object::Tap   656455/s      1%          --        -5%    -13%         -29%   -48%
    tap-reftap    688073/s      6%          5%         --     -9%         -26%   -46%
    own-tap       755668/s     16%         15%        10%      --         -18%   -40%
    tap-autoload  925926/s     42%         41%        35%     23%           --   -27%
    no-tap       1265823/s     95%         93%        84%     68%          37%     --
    perl -l ex/bench.pl  26.30s user 0.04s system 99% cpu 26.362 total

As mentioned earlier, the fastest case `no-tap` is just not using tap - but
where's the fun in that?

The three slowest cases are ["tap::tap"](#tap-tap), [Object::Tap](https://metacpan.org/pod/Object::Tap), and saving
a reference to ["tap::tap"](#tap-tap), similar to what you get with Object::Tap.  The
next case, `own-tap`, is just the object having its own tap method, which
should also be similar to having it in a parent class.  `tap-autoload` is
["tap::\*"](#tap), which edges the others out by making the `tap` package its own
sort of method lookup cache.  The trade-off is a package that grows at runtime.

I'm open to any other ideas to do even better, I've considered some myself, but
it's also not bad to just avoid using tap in very hot code.

# AUTHOR

Meredith Howard <mhoward@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2019 by Meredith Howard.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
0.01 2019-04-19T04:31:08
 - init Changes (Meredith Howard)
 - Docs (Meredith Howard)
 - Update README.md (Meredith Howard)
 - Add perf section (Meredith Howard)
 - Add own-tap case (Meredith Howard)
 - Lower min perl requirement to 5.6.0 (Meredith Howard)
 - Show the alternative for clarity (Meredith Howard)
 - Fix POD and README.md (Meredith Howard)
 - Add README (Meredith Howard)
 - Initial commit (Meredith Howard)
 - initialize (Meredith Howard)


t/Foo.pm
package Foo;

sub new { bless {}, shift }

sub color {
  my ($self, $color) = @_;
  $self->{color} = $color;
}

sub finish {
  my ($self, $finish) = @_;
  $self->{finish} = $finish;
}

sub describe {
  my ($self) = @_;
  sprintf "Why, it's a %s, %s Foo!",
    @{$self}{'finish', 'color'};
}

1;


ex/bench.pl
#!/usr/bin/env perl

use warnings;
use strict;
use 5.010;

use FindBin;
use lib "$FindBin::Bin/../lib";

use tap;
use Object::Tap qw($_tap);

use Benchmark qw(cmpthese);

package Foo {
  sub new { bless {}, shift }
  sub color {
    $_[0]->{color} = $_[1];
  }
  sub finish {
    $_[0]->{finish} = $_[1];
  }
  sub tap {
    my ($self, $method) = (shift, shift);
    $self->$method(@_);
    $self;
  }
};

# use DDP;
# Foo->new
#   ->tap::color('pink')
#   ->tap::finish('matte')
#   ->tap::tap(\&p);

cmpthese 3_000_000, {
  'tap-autoload' => sub {
    Foo->new
      ->tap::color('pink')
      ->tap::finish('matte')
  },
  'tap-tap' => sub {
    Foo->new
      ->tap::tap(color => 'pink')
      ->tap::tap(finish => 'matte')
  },
  'tap-reftap' => sub {
    state $tap = \&tap::tap;
    Foo->new
      ->$tap(color => 'pink')
      ->$tap(finish => 'matte')
  },
  'Object::Tap' => sub {
    Foo->new
      ->$_tap(color => 'pink')
      ->$_tap(finish => 'matte')
  },
  'no-tap' => sub {
    my $foo = Foo->new;
    $foo->color('pink');
    $foo->finish('matte');
    $foo
  },
  'own-tap' => sub {
    Foo->new
      ->tap(color => 'pink')
      ->tap(finish => 'matte')
  },
};


lib/tap.pm
package tap;
# ABSTRACT: add a tap method everywhere

use warnings;
use strict;

use vars qw($AUTOLOAD);
use constant PREFIX_LENGTH => length('tap::');

=head1 SYNOPSIS

  use tap;

  return Some::Class->new
    ->tap::color('red')
    ->tap::smell('roses');

  # rather than:

  my $o = Some::Class->new;
  $o->color('red')
  $o->smell('roses');
  return $o;

=head1 DESCRIPTION

C<tap> is a method that wraps any other method, ignoring its result and
returning the invocant instead.  This allows you to make chaining method out of
anything, such a setters that return what you just set, or to peek into
a method chain.

This module gets you access to tap anywhere, without requiring it to exist in
a parent class or an imported lexical.

=head1 METHODS

=head2 tap::*

  $o->tap::color('red')
    ->tap::smell('roses')
    ->...;

For use when you know what existing method you want to call.  This is
implemented using AUTOLOAD, which follows the common behavior of installing
a method so you only fall to AUTOLOAD once.

It's not as fast as writing things out the long way, but it's the fastest of
all tap-like features I've seen.  And I think it's convenient while looking
nice.  I was using it in L<Reply> for quite a while before making it a module.

=cut

sub AUTOLOAD {
  my ($self, @args) = @_;
  my $method = substr($AUTOLOAD, PREFIX_LENGTH())
    or return;

  eval sprintf(<<'END_PERL', $method) unless $method eq 'tap';
    sub %1$s {
      my $self = shift;
      $self->%1$s(@_);
      $self;
    }
END_PERL

  $self->$method(@args);
  $self;
}

=head2 tap::tap

  $o->tap::tap($method => 42)->...;

  # slip debugging into a chain
  use DDP;
  $o->tap::tap(\&p)->run;
 
For when the method name is dynamic, or you want to pass a coderef to be called
as a method.  For convenience the invocant is topicalized (C<$_>) as well:

  $o->tap::tap(sub { $_->foo->bar(42) })->run;

=cut

sub tap {
  my ($self, $method) = (shift, shift);
  $self->$method(@_) for $self;
  $self;
}

=head1 SEE ALSO

L<Object::Tap>, L<Object::Util>, L<Mojo::Base> - other taps

L<curry> - interface inspiration

=head1 PERFORMANCE

The following example benchmark (C<ex/bench.pl>) was run with perl 5.26.0 on
a 2015 MacBook Pro.

                    Rate tap-tap Object::Tap tap-reftap own-tap tap-autoload no-tap
  tap-tap       650759/s      --         -1%        -5%    -14%         -30%   -49%
  Object::Tap   656455/s      1%          --        -5%    -13%         -29%   -48%
  tap-reftap    688073/s      6%          5%         --     -9%         -26%   -46%
  own-tap       755668/s     16%         15%        10%      --         -18%   -40%
  tap-autoload  925926/s     42%         41%        35%     23%           --   -27%
  no-tap       1265823/s     95%         93%        84%     68%          37%     --
  perl -l ex/bench.pl  26.30s user 0.04s system 99% cpu 26.362 total

As mentioned earlier, the fastest case C<no-tap> is just not using tap - but
where's the fun in that?

The three slowest cases are L</tap::tap>, L<Object::Tap>, and saving
a reference to L</tap::tap>, similar to what you get with Object::Tap.  The
next case, C<own-tap>, is just the object having its own tap method, which
should also be similar to having it in a parent class.  C<tap-autoload> is
L</tap::*>, which edges the others out by making the C<tap> package its own
sort of method lookup cache.  The trade-off is a package that grows at runtime.

I'm open to any other ideas to do even better, I've considered some myself, but
it's also not bad to just avoid using tap in very hot code.

=cut

1;
